src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -380,79 +382,254 @@ internal static BigInteger ParseBigInteger(ReadOnlySpan<char> value, NumberStyle
-            if (!TryParseBigInteger(value, style, info, out result))
+            if (!TryParseBigInteger(value, style, info, out BigInteger result))
             {
                 throw new FormatException(SR.Overflow_ParseBigInteger);
             }
             return result;
         }
 
-        private static unsafe bool HexNumberToBigInteger(ref BigNumberBuffer number, ref BigInteger value)
+        private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInteger result)
         {
             if (number.digits == null || number.digits.Length == 0)
+            {
+                result = default;
                 return false;
+            }
+
+            const int DigitsPerBlock = 8;
+
+            int totalDigitCount = number.digits.Length - 1;   // Ignore trailing '\0'
+            int blockCount, partialDigitCount;
 
-            int len = number.digits.Length - 1; // Ignore trailing '\0'
-            byte[] bits = new byte[(len / 2) + (len % 2)];
+            if ((uint)totalDigitCount % DigitsPerBlock == 0)
+            {
+                blockCount = totalDigitCount / DigitsPerBlock;
+                partialDigitCount = 0;
+            }
+            else
+            {
+                blockCount = Math.DivRem(totalDigitCount, DigitsPerBlock, out int mod) + 1;
+                partialDigitCount = DigitsPerBlock - mod;
+            }
+
+            bool isNegative = HexConverter.FromChar(number.digits[0]) >= 8;
+            uint partialValue = (isNegative && partialDigitCount > 0) ? 0xFFFFFFFFu : 0;
+
+            int[]? arrayFromPool = null;
 
-            bool shift = false;
-            bool isNegative = false;
-            int bitIndex = 0;
+            Span<int> bitsBuffer = (blockCount <= BigInteger.StackallocUInt32Limit)
+                ? stackalloc int[blockCount]
+                : (arrayFromPool = ArrayPool<int>.Shared.Rent(blockCount)).AsSpan(0, blockCount);
 
-            // Parse the string into a little-endian two's complement byte array
-            // string value     : O F E B 7 \0
-            // string index (i) : 0 1 2 3 4 5 <--
-            // byte[] (bitIndex): 2 1 1 0 0 <--
-            //
-            for (int i = len - 1; i > -1; i--)
+            int bitsBufferPos = blockCount - 1;
+
+            try
             {
-                char c = number.digits[i];
-                int b = HexConverter.FromChar(c);
-                Debug.Assert(b != 0xFF);
-                if (i == 0 && (b & 0x08) == 0x08)
-                    isNegative = true;
+                foreach (ReadOnlyMemory<char> digitsChunkMem in number.digits.GetChunks())
+                {
+                    ReadOnlySpan<char> chunkDigits = digitsChunkMem.Span;
+                    for (int i = 0; i < chunkDigits.Length; i++)
+                    {
+                        char digitChar = chunkDigits[i];
+                        if (digitChar == '\0')
+                            break;
```

Comment:
nit: I strongly prefer always using braces to help reduce refactoring and reading bugs:
```suggestion
                        if (digitChar == '\0')
                        {
                            break;
                        }
```

Reply:
This looks to apply to a few changes in the PR